ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -96,6 +96,10 @@ CompletableFuture<Long> write(ByteBuf buf, Iterable<WriteOption> options,
           n -> streamFuture.thenCompose(stream -> writeToAsync(buf, options, stream, executor)
               .whenComplete((l, e) -> metrics.stop(context, e == null))));
     }
+
+    void cleanUp() {
+      streamFuture.thenAccept(DataStream::cleanUp);
+    }
```

Comment:
hi, @szetszwo 
Whether the current implementation of ozone throws IOE if cleanUp is called 

see:
https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/LocalStream.java#L46

https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java#L246

Reply:
When there is an IOE, it should still close the file as below.
```java
+++ b/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java
@@ -198,8 +198,11 @@ void assertOpen() throws IOException {
   @Override
   public void close() throws IOException {
     if (closed.compareAndSet(false, true)) {
-      putBlockRequest.set(closeBuffers(buffers, super::writeFileChannel));
-      super.close();
+      try {
+        putBlockRequest.set(closeBuffers(buffers, super::writeFileChannel));
+      } finally {
+        super.close();
+      }
     }
   }
```
Also, it probably should also delete the file
```java
+++ b/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/LocalStream.java
@@ -46,6 +46,7 @@ public CompletableFuture<?> cleanUp() {
         dataChannel.close();
         return true;
       } catch (IOException e) {
+        // TODO delete the file?
         throw new CompletionException("Failed to close data channel", e);
       }
     });
```

Reply:
@guohao-rosicky , please test it with https://issues.apache.org/jira/secure/attachment/13058854/cleanup.patch .  It will also delete the file.

Reply:
hi, @szetszwo .

Is it safer to delete files after the container is closed and then scan the data in the container in the background service?

see:
https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java#L72

Reply:
> Is it safer to delete files after the container is closed and then scan the data in the container in the background service?

@guohao-rosicky , it is a good idea.  Then, we mostly need try-finally (the other changes are for null checks); see https://issues.apache.org/jira/secure/attachment/13058865/try-finally.patch